File: vtable-example.cpp
#include <iostream>
using namespace std;

class Draw
{
    public:
      virtual void draw();
};

class Button : public Draw
{
    public:
       void draw()
         { this->Draw::draw();
           cout << "Drawing button.\n"; 
         }
};

int main()
{
    Draw *w;
    Button b;

    w = &b;
    w->draw();

    return 0;
}

File: vtable-example.rs
// Type your code here, or load an example.
pub trait Draw {
    fn draw(&self);

    // what happens when we uncomment this?
    // fn draw2(&self){}
}

pub struct Button;

impl Draw for Button {
    fn draw(&self) {
        println!("Drawing button ...");
    }
}

fn shoot(w: &dyn Draw) {
    w.draw();
}

pub fn do_it() {
    shoot(&Button{});
}